labgrid/driver/shelldriver.py

```diff
@@ -71,39 +91,151 @@ def on_activate(self):
         self._run("dmesg -n 1")
 
     def on_deactivate(self):
+        if self.background_process.running:
+            self.console.sendcontrol('c')
+            self.console.expect(self.prompt)
+            self.background_process.exitcode = 130
+
         self._status = 0
 
+    @Driver.check_active
+    @step(args=['cmd'], result=True)
+    def run(self, cmd, timeout=30.0, codec="utf-8", decodeerrors="strict"):
+        return self._run(cmd, timeout=timeout, codec=codec, decodeerrors=decodeerrors)
+
     def _run(self, cmd, *, timeout=30.0, codec="utf-8", decodeerrors="strict"):
         """
         Runs the specified cmd on the shell and returns the output.
 
         Arguments:
         cmd - cmd to run on the shell
         """
-        # FIXME: Handle pexpect Timeout
+        timeout_ = None if timeout is None else Timeout(float(timeout))
+
+        process = self._run_as_background_process(cmd)
+        read_timeout = timeout_.remaining if timeout_ else None
+        stdout, stderr = self._read_from_background_process(process, timeout=read_timeout)
+
+        if timeout_ and timeout_.expired:
```

Comment:
Maybe this is a race condition? Timeout can expire short after _read_from_background_process successfully read till exitcode. 
```suggestion
        if timeout_ and timeout_.expired and self.background_process.running:
```
Do we need to cleanup the process in case of a timeout here?

Reply:
You are right, this situation can happen. Here in `ShellDriver._run()` your solution is correct. But what about the usage example of the _BackgroundProcessProtocol_ [here](https://github.com/labgrid-project/labgrid/pull/552/files#diff-8032c949da39d97978728cb39890d762R12)? Should we live this corner case there?

Reply:
Regarding clean up: yes, we need to clean up in that case, that's missing. I guess we should move the background process clean up lines from `ShellDriver.on_deactivate()` into a separate function. Users of the BackgroundProcessProtocol need to do this themselves too, so we should add something like `BackgroundProcessProtocol.clean_up_background_process()`. I am still not sure how to expose all this is in a user-friendly way. Suggestions welcome :)

Reply:
My usecase is to start a not stopping command (like iperf -s) on the target, send some data over other interface and than force stop the command on the target and get the output so far.
So I like the idea of `BackgroundProcessProtocol.clean_up_background_process()`, maybe we call it `BackgroundProcessProtocol.stop_background_process()`

Reply:
for the usage example of the BackgroundProcessProtocol: whats about adding the timout to `BackgroundProcessProtocol.run_as_background_process`? `BackgroundProcessProtocol.poll_background_process()` could than raise TIMEOUT if needed

something like
```
process = run_as_background_process('dd if=/dev/random bs=1M count=100', timeout=300.0)
try:
    while driver.poll_background_process(process) is None:
       stdout, stderr = driver.read_from_background_process(process, timeout=1.0)
       # do something else
    exitcode = driver.poll_background_process(process)
  except TIMEOUT:
      driver.stop_background_process(process)
```
`BackgroundProcessProtocol.read_from_background_process()` blocks here and servers as sleep.